drivers/power/adp1050/adp1050.c

```diff
@@ -0,0 +1,1252 @@
+ * @param pass_type - Password type to be used for register lockage.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int adp1050_lock_pass(struct adp1050_desc *desc, uint16_t password,
+		      enum adp1050_pass_type pass_type)
+{
+	switch (pass_type) {
+	case ADP1050_CHIP_PASS:
+		return adp1050_write(desc, ADP1050_CHIP_PASSWORD, password, 2);
+	case ADP1050_EEPROM_PASS:
+		return adp1050_write(desc, ADP1050_EEPROM_PASSWORD,
+				     password, 1);
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief Change the CHIP/EEPROM password. TRIM password is the same with
+ * 	  EEPROM password.
+ * @param desc - ADP1050 device descriptor
+ * @param old_pass - Old password value.
+ * @param new_pass - New password value.
+ * @param pass_type - Password type (CHIP/EEPROM).
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int adp1050_change_pass(struct adp1050_desc *desc, uint16_t old_pass,
+			uint16_t new_pass, enum adp1050_pass_type pass_type)
+{
+	int ret;
+
+	if (pass_type == ADP1050_EEPROM_PASS && new_pass > ADP1050_EEPROM_DEFAULT_PASS)
+		return -EINVAL;
+
+	switch (pass_type) {
+	case ADP1050_CHIP_PASS:
+		ret = adp1050_write(desc, ADP1050_CHIP_PASSWORD, old_pass, 2);
+		if (ret)
+			return ret;
+
+		ret = adp1050_write(desc, ADP1050_CHIP_PASSWORD, new_pass, 2);
+		if (ret)
+			return ret;
+
+		return adp1050_send_command(desc, ADP1050_STORE_USER_ALL);
+	case ADP1050_EEPROM_PASS:
+		ret = adp1050_write(desc, ADP1050_EEPROM_PASSWORD, old_pass, 1);
+		if (ret)
+			return ret;
+
+		return adp1050_write(desc, ADP1050_EEPROM_PASSWORD,
+				     new_pass, 1);
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief Software reset the ADP1050.
+ * @param desc - ADP1050 device descriptor
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int adp1050_software_reset(struct adp1050_desc *desc)
+{
+	int ret;
+
+	ret = adp1050_write(desc, ADP1050_SOFTWARE_RESET_SETTINGS,
+			    ADP1050_SW_RES_DELAY_500MS
+			    | ADP1050_SW_RES_NO_DELAY, 1);
+	if (ret)
+		return ret;
+
+	return adp1050_write(desc, ADP1050_SOFTWARE_RESET_GO,
+			     ADP1050_SOFTWARE_RESET_ON, 1);
+}
+
+/**
+ * @brief Initialize the ADP1050 device.
+ * @param desc - ADP1050 device descriptor
+ * @param init_param - Initialization parameter containing information about the
+ * 		       ADP1050 device to be initialized.
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int adp1050_init(struct adp1050_desc **desc,
+		 struct adp1050_init_param *init_param)
+{
+	struct adp1050_desc *descriptor;
+	int ret;
+
+	descriptor = (struct adp1050_desc *)no_os_calloc(sizeof(*descriptor), 1);
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = no_os_i2c_init(&descriptor->i2c_desc, init_param->i2c_param);
+	if (ret) {
+		no_os_free(descriptor);
+		return ret;
+	}
+
+	ret = no_os_gpio_get_optional(&descriptor->pg_alt_desc,
+				      init_param->pg_alt_param);
+	if (ret)
+		return ret;
```

Comment:
please free descriptor before returning, and remove i2c driver.... 
please treat memory leaks in the whole API, no_os_gpio_get_optional allocates memory as well, which should be freed in case of an error, before returning.